File: mcp-server/internal/domains/mcpservers/fetch.go
// Package mcpservers provides the MCP tools and resource template for the
// McpServer domain, backed by the McpServerQueryController and
// McpServerCommandController RPCs.
package mcpservers

import (
	"context"
	"fmt"

	"github.com/stigmer/stigmer/mcp-server/internal/domains"
	mcpserverv1 "github.com/stigmer/stigmer/mcp-server/proto/ai/stigmer/agentic/mcpserver/v1"
	"github.com/stigmer/stigmer/mcp-server/proto/ai/stigmer/commons/apiresource"
	"github.com/stigmer/stigmer/mcp-server/proto/ai/stigmer/commons/apiresource/apiresourcekind"
	"google.golang.org/grpc"
)

// Fetch retrieves an MCP server by org and slug, returning its JSON
// representation. Both tool and resource handlers delegate to this function
// for the actual RPC.
func Fetch(ctx context.Context, serverAddress, org, slug string) (string, error) {
	return domains.WithConnection(ctx, serverAddress,
		func(ctx context.Context, conn *grpc.ClientConn) (string, error) {
			client := mcpserverv1.NewMcpServerQueryControllerClient(conn)
			mcpServer, err := client.GetByReference(ctx, &apiresource.ApiResourceReference{
				Org:  org,
				Kind: apiresourcekind.ApiResourceKind_mcp_server,
				Slug: slug,
			})
			if err != nil {
				return "", domains.RPCError(err, fmt.Sprintf("MCP server %q in org %q", slug, org))
			}
			return domains.MarshalJSON(mcpServer)
		})
}

File: mcp-server/src/domains/mcpservers/fetch.ts
// MCP server read path: the single RPC both the get_mcp_server tool and the
// mcp-server resource template delegate to.
// Go parity: mcp-server/internal/domains/mcpservers/fetch.go.

import { McpServerSchema } from "@stigmer/protos/ai/stigmer/agentic/mcpserver/v1/api_pb";
import { McpServerQueryController } from "@stigmer/protos/ai/stigmer/agentic/mcpserver/v1/query_pb";
import { ApiResourceKind } from "@stigmer/protos/ai/stigmer/commons/apiresource/apiresourcekind/api_resource_kind_pb";

import { withClient } from "../client.js";
import { toProtoJson } from "../marshal.js";
import { rpcError } from "../rpcerr.js";

/**
 * Retrieve an MCP server by org and slug, returning its protojson
 * representation. Errors are classified into user-facing messages via
 * {@link rpcError}.
 */
export async function fetchMcpServer(
  serverAddress: string,
  token: string,
  org: string,
  slug: string,
): Promise<string> {
  return withClient(McpServerQueryController, serverAddress, token, async (client, callOptions) => {
    try {
      const mcpServer = await client.getByReference(
        { org, kind: ApiResourceKind.mcp_server, slug },
        callOptions,
      );
      return toProtoJson(McpServerSchema, mcpServer);
    } catch (err) {
      throw rpcError(err, `MCP server "${slug}" in org "${org}"`);
    }
  });
}
